src/core/lib/promise/detail/promise_like.h

```diff
@@ -71,7 +71,12 @@ class PromiseLike<void>;
 
 template <typename F>
 class PromiseLike<F, absl::enable_if_t<!std::is_void<
-                         typename std::result_of<F()>::type>::value>> {
+#if __cplusplus >= 201703L
```

Comment:
we probably need this to be the same as what absl uses - 
```
#if (defined([__cpp_lib_is_invocable](https://source.corp.google.com/piper///depot/google3/third_party/absl/meta/type_traits.h;bpv=1;bpt=1;l=324?q=absl::type_traits_internal::result_of&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental&gsn=__cpp_lib_is_invocable&gs=KYTHE%3A%2F%2FItUECpUBa3l0aGU6Ly9nb29nbGUzP2xhbmc9YyUyQiUyQj9wYXRoPXRoaXJkX3BhcnR5L2FwcGxlX3Nka3MveGNvZGVfMTVfMS9pcGhvbmVzaW11bGF0b3IvdXNyL2luY2x1ZGUvYyUyQiUyQi92MS92ZXJzaW9uI19fY3BwX2xpYl9pc19pbnZvY2FibGUlMjNtJTQwMTgwODMKjAFreXRoZTovL2dvb2dsZTM_bGFuZz1jJTJCJTJCP3BhdGg9dGhpcmRfcGFydHkvYXBwbGVfc2Rrcy94Y29kZV8xNV8xL21hY29zeC91c3IvaW5jbHVkZS9jJTJCJTJCL3YxL3ZlcnNpb24jX19jcHBfbGliX2lzX2ludm9jYWJsZSUyM20lNDAxODA4MwqOAWt5dGhlOi8vZ29vZ2xlMz9sYW5nPWMlMkIlMkI_cGF0aD10aGlyZF9wYXJ0eS9jcm9zc3Rvb2wvdjE4L3N0YWJsZS9pbnN0YWxscy9sbHZtL2luY2x1ZGUvYyUyQiUyQi92MS92ZXJzaW9uI19fY3BwX2xpYl9pc19pbnZvY2FibGUlMjNtJTQwMjE0MTgKmgFreXRoZTovL2dvb2dsZTM_bGFuZz1jJTJCJTJCP3BhdGg9dGhpcmRfcGFydHkvdW5zdXBwb3J0ZWRfdG9vbGNoYWlucy9mdWNoc2lhL3Rvb2xjaGFpbi9jbGFuZy9pbmNsdWRlL2MlMkIlMkIvdjEvdmVyc2lvbiNfX2NwcF9saWJfaXNfaW52b2NhYmxlJTIzbSU0MDIxNDE4)) && [__cpp_lib_is_invocable](https://source.corp.google.com/piper///depot/google3/third_party/absl/meta/type_traits.h;bpv=1;bpt=1;l=324?q=absl::type_traits_internal::result_of&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental&gsn=__cpp_lib_is_invocable&gs=KYTHE%3A%2F%2FItUECpUBa3l0aGU6Ly9nb29nbGUzP2xhbmc9YyUyQiUyQj9wYXRoPXRoaXJkX3BhcnR5L2FwcGxlX3Nka3MveGNvZGVfMTVfMS9pcGhvbmVzaW11bGF0b3IvdXNyL2luY2x1ZGUvYyUyQiUyQi92MS92ZXJzaW9uI19fY3BwX2xpYl9pc19pbnZvY2FibGUlMjNtJTQwMTgwODMKjAFreXRoZTovL2dvb2dsZTM_bGFuZz1jJTJCJTJCP3BhdGg9dGhpcmRfcGFydHkvYXBwbGVfc2Rrcy94Y29kZV8xNV8xL21hY29zeC91c3IvaW5jbHVkZS9jJTJCJTJCL3YxL3ZlcnNpb24jX19jcHBfbGliX2lzX2ludm9jYWJsZSUyM20lNDAxODA4MwqOAWt5dGhlOi8vZ29vZ2xlMz9sYW5nPWMlMkIlMkI_cGF0aD10aGlyZF9wYXJ0eS9jcm9zc3Rvb2wvdjE4L3N0YWJsZS9pbnN0YWxscy9sbHZtL2luY2x1ZGUvYyUyQiUyQi92MS92ZXJzaW9uI19fY3BwX2xpYl9pc19pbnZvY2FibGUlMjNtJTQwMjE0MTgKmgFreXRoZTovL2dvb2dsZTM_bGFuZz1jJTJCJTJCP3BhdGg9dGhpcmRfcGFydHkvdW5zdXBwb3J0ZWRfdG9vbGNoYWlucy9mdWNoc2lhL3Rvb2xjaGFpbi9jbGFuZy9pbmNsdWRlL2MlMkIlMkIvdjEvdmVyc2lvbiNfX2NwcF9saWJfaXNfaW52b2NhYmxlJTIzbSU0MDIxNDE4) >= 201703L) || \
    (defined(_MSVC_LANG) && _MSVC_LANG >= 201703L)
    ```

Reply:
Yeah, that make sense. I'll update it.

I wonder why abseil has the separate `defined(_MSVC_LANG) && _MSVC_LANG >= 201703L` check though. I do see that MSVC leaves `__cplusplus` defined to an old version, but it does appear to defined `__cpp_lib_is_invocable` to a new value.